### PR TITLE
Implement Distinct On for Postgres

### PR DESCRIPTION
**Pull request by weiznich**
Hopefully I put everything in the correct place.

(It is not possible to implement this outside of diesel because of them `DistinctOnDsl` impl for `SelectStatement`)

### CODE REVIEW

File: diesel/src/query_builder/select_statement/dsl_impls.rs [hunk 1]
```diff
@@ -86,6 +86,36 @@ where
     }
 }
 
+#[cfg(feature = "postgres")]
```
Comment (killercup) [inline diesel/src/query_builder/select_statement/dsl_impls.rs]: Not sure this (the generic `query_builder` mod) is the right place to a postgres specific impl?

File: diesel/src/query_dsl/distinct_dsl.rs [hunk 2]
```diff
@@ -46,3 +47,55 @@ where
         self.as_query().distinct()
     }
 }
+
+/// Adds the `DISTINCT ON` clause to a query.
+///
+/// # Example
+///
+/// ```rust
+///
```
Comment (killercup) [inline diesel/src/query_dsl/distinct_dsl.rs]: remove empty line

File: diesel/src/query_dsl/distinct_dsl.rs [hunk 3]
```diff
@@ -46,3 +47,55 @@ where
...
+/// #         id -> Integer,
+/// #         name -> VarChar,
+/// #     }
+/// # }
+/// #
+/// # fn main() {
+/// #     use self::animals::dsl::*;
+/// #     let connection = establish_connection();
+/// #     connection.execute("DELETE FROM animals").unwrap();
+/// connection.execute("INSERT INTO animals (species, name, legs) VALUES ('dog', 'Jack', 4), ('dog', Null, 4), ('spider', Null, 8)")
```
Comment (killercup) [inline diesel/src/query_dsl/distinct_dsl.rs]: maybe split that string at VALUES into two lines

File: diesel/src/query_dsl/distinct_dsl.rs [hunk 4]
```diff
@@ -46,3 +47,55 @@ where
...
+///     .unwrap();
+/// let all_animals = animals.select((species, name, legs)).load(&connection);
+/// let distinct_animals = animals.select((species, name, legs)).distinct_on(species).load(&connection);
+///
+/// assert_eq!(Ok(vec![(String::from("dog"), Some(String::from("Jack")), 4),
+///                    (String::from("dog"), None, 4),
+///                    (String::from("spider"), None, 8)]), all_animals);
+/// assert_eq!(Ok(vec![(String::from("dog"), Some(String::from("Jack")), 4),
+///                    (String::from("spider"), None, 8)]), distinct_animals);
```
Comment (killercup) [inline diesel/src/query_dsl/distinct_dsl.rs]: This is hard to parse for me. Maybe swap the params so the var is first? Or make a small `Animal::new("dog", Some("Jack"), 4)` builder (i.e., `fn new<S: Into<String>, N: Into<String>>(species: S, name: Option<N>, legs: u32) { … }` to reduce the `String::from` noise)

File: diesel/src/query_dsl/distinct_dsl.rs [hunk 5]
```diff
@@ -1,4 +1,5 @@
 use query_source::Table;
+use expression::SelectableExpression;
```
Comment (killercup) [inline diesel/src/query_dsl/distinct_dsl.rs]: sorry :(

```
error: unused import: `expression::SelectableExpression`
 --> src/query_dsl/distinct_dsl.rs:2:5
  |
2 | use expression::SelectableExpression;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: lint level defined here
 --> src/lib.rs:17:9
  |
17| #![deny(warnings, missing_debug_implementations, missing_copy_implementations)]
  |         ^^^^^^^^
  = note: #[deny(unused_imports)] implied by #[deny(warnings)]
```

File: diesel/src/query_builder/select_statement/dsl_impls.rs (hunk 1 above)
Comment (weiznich) [reply diesel/src/query_builder/select_statement/dsl_impls.rs]: So better move this to pg/query_builder/query_fragment_impl.rs?

Comment (killercup) [reply diesel/src/query_builder/select_statement/dsl_impls.rs]: Yeah, but maybe someone else has a good idea for how to organize these impls?

**Comment by sgrif** [comment]
> (It is not possible to implement this outside of diesel because of them DistinctOnDsl impl for SelectStatement)

That doesn't sound right. Assuming `DistinctOnDsl` comes from your crate, you should be able to implement it for any type. (Note: I'm not opposed to having this in Diesel)

**Comment by weiznich** [comment]
> That doesn't sound right. Assuming `DistinctOnDsl` comes from your crate, you should be able to implement it for any type. (Note: I'm not opposed to having this in Diesel)

The problem is not that I cannot implement `DistinctOnDsl` for `SelectStatement` due to the orphan rule, but because I cannot access the fields of `SelectStatement` to construct the new version of `SelectStatement`.

File: diesel/src/query_builder/select_statement/mod.rs [hunk 6]
```diff
@@ -43,15 +43,15 @@ pub struct SelectStatement<
...
-    for_update: ForUpdate,
+    pub(crate) select: Select,
+    pub(crate) from: From,
+    pub(crate) distinct: Distinct,
+    pub(crate) where_clause: Where,
+    pub(crate) order: Order,
+    pub(crate) limit: Limit,
+    pub(crate) offset: Offset,
+    pub(crate) group_by: GroupBy,
+    pub(crate) for_update: ForUpdate,
```
Comment (killercup) [inline diesel/src/query_builder/select_statement/mod.rs]: Probably a good solution

**Comment by sgrif** [comment]
LGTM

**Comment by killercup** [comment]
Thanks!
